3Sum.py

```diff
@@ -1,19 +1,23 @@
 class Solution:
-  def threeSum(self, nums: List[int]) -> List[List[int]]:
-    result = set()
-    c = Counter(nums)
-    nums = []
-    for k, v in c.most_common():
-      nums.extend([k, k] if v > 1 else [k])
-      
-    for i, n1 in enumerate(nums):
-      c[n1] -= 1
-      for n2 in nums[i+1:]:
-        n3 = -(n1 + n2)
-        c[n2] -= 1
-        if n3 in c and c[n3] > 0:
-          numbs = tuple(sorted([n1, n2, n3]))
-          result.add(numbs)
-        c[n2] += 1
-    print(nums)
-    return result
\ No newline at end of file
+    def threeSum(self, nums: List[int]) -> List[List[int]]:
+        result = set()  # Will save only unique arrays
+        # we need to find a + b + c == 0, that is same as a + b == -c, so we will use that to check for -c
+        c = Counter(nums)
+        nums = []  # Counted all elements and cleaned array
+        for k, v in c.most_common():  # Getting number and amount of that number in array
+            # Removing more than two occurrences of same number (so will not exceed time limit)
+            nums.extend([k, k] if v > 1 else [k])
+
+        for i, n1 in enumerate(nums):  # enumerate returns index, value of list
+            c[n1] -= 1  # Removing used numbers from counted nums
+            for n2 in nums[i+1:]:
+                n3 = -(n1 + n2)
+                c[n2] -= 1  # removing used number
+                if n3 in c and c[n3] > 0:
+                    # sorting and using tuples, so we will have unique result
+                    numbs = tuple(sorted([n1, n2, n3]))
+                    result.add(numbs)
+                # adding back used number, becuase we will use it in further itarations
+                c[n2] += 1
```

Comment:
I don't see where `c[n2]` is being used right after we subtract 1 (and before we get it back here) 🤔

Reply:
if used on line 16: 
```if n3 in c and c[n3] > 0:```
If we will not return used number, after first iteration we will have no numbers in dict, and nothing to check with

Reply:
I don't see a relation between `n2` & `n3`, probably that's the reason I couldn't reason about that line. So, is it like `c2` could equal `c3`?

Reply:
We need to find combination of n1, n2, n3, that `n1 + n2 + n3 == 0`, but that equal to `n1 + n2 == - n3`.
So we pick random n1 and n2, remove them from dict (because we can not use them anymore), and search for -n3 in dict.
 After we finish, we discard n1, because we don't want to any combination use this number but return return n2, because n2 may be used by next n1. (So, we basically search for every combination of 2 el in list)